### ISSUE TITLE
About Layernorms within MSAB

### ISSUE THREAD

**Issue by yyyang404** [caiyuanhao1998/MST#28]
```
class MSAB(nn.Module):
    def __init__(
            self,
            dim,
            dim_head=64,
            heads=8,
            num_blocks=2,
    ):
        super().__init__()
        self.blocks = nn.ModuleList([])
        for _ in range(num_blocks):
            self.blocks.append(nn.ModuleList([
                MS_MSA(dim=dim, dim_head=dim_head, heads=heads),
                PreNorm(dim, FeedForward(dim=dim))
            ]))

    def forward(self, x):
        x = x.permute(0, 2, 3, 1)
        for (attn, ff) in self.blocks:
            x = attn(x) + x
            x = ff(x) + x
        out = x.permute(0, 3, 1, 2)
        return out
```

I wonder if there is a Layernorm missing in MSAB, since the MST paper describes "MSAB is composed of a Feed-Forward Network (FFN), an MS-MSA, and two layer normalization"

**Comment by caiyuanhao1998** [comment]
Hi, 

Thanks for your interests. Another layer normalization is in Line 150-151 of MS_MSA module in MST.py, check [this](https://github.com/caiyuanhao1998/MST/blob/main/simulation/train_code/architecture/MST.py#:~:text=q%20%3D%20F.normalize(q%2C%20dim,dim%3D%2D1%2C%20p%3D2))

**Comment by yyyang404** [comment]
Thanks for your quick reply and i hope that won't bother you.
So the normalize is only applied on Q and K in computing attention matrix, leaving V not normed. That's different from the LN in common transformer-based models which applies to the whole feature map. I wonder what the reason this design is for?

**Comment by caiyuanhao1998** [comment]
we aim to stabilize the spectral-wise self-attention map produced by Q and K. For V, we apply mask-attention and use LN to process it in the subsequent layers. I can also try to apply an LN to process V in the MS-MSA module.

**Comment by yyyang404** [comment]
okay i got it. appreciate your great work :)
